R/base.R

```diff
@@ -35,8 +31,8 @@ base_default <- function(type) {
     list = list(),
     expression = expression(),
```

Comment:
>  .... because R won't display a argument with default `1:10` and `quote(1:10)` differently. 

Huh?  not really .. (I assume you meant something else...) :
```
> ff <- function(x = 1:10, y = quote(1:10)) { x - eval(y) }
> ff()
 [1] 0 0 0 0 0 0 0 0 0 0
> ff
function(x = 1:10, y = quote(1:10)) { x - eval(y) }
> args(ff)
function (x = 1:10, y = quote(1:10)) 
NULL
> 
```

Reply:
I meant this problem:

``` r
f <- function(x, y) {}
formals(f) <- list(x = 1:10, y = quote(1:10))
f
#> function (x = 1:10, y = 1:10) 
#> {
#> }
```

<sup>Created on 2024-09-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Reply:
I think this might be what @hadley is describing (please confirm) 
``` r
# literal constants get inlined
f <- function(x = 1) x
typeof(formals(f)$x)
#> [1] "double"

# expressions stay expressions in 'normally' defined functions
f <- function(x = 1:10) x
typeof(formals(f)$x)
#> [1] "language"
f
#> function(x = 1:10) x

# but, generated functions with arguments that cannot be generated 
# as a literal by the parser present the same as "normally" defined functions
formals(f)$x <- 1:10
typeof(formals(f)$x)
#> [1] "integer"
f # prints the same as before, even though x is no longer type 'language'
#> function (x = 1:10) 
#> x
```

<sup>Created on 2024-09-09 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
I missed @hadley's last comment before posting mine.

Same issue, however:
``` r
f1 <- f2 <- function(x = 1:10) {}
formals(f1) <- list(x = quote(1:10))
identical(f1, f2)
#> [1] TRUE
```

Reply:
Back to the point, I think it's mostly a non-issue for object types that don't have reference semantics. The only reason to quote `new.env()` is that, otherwise, all instances that define a property of that type would default to sharing the same environment. `expression()` does not have reference semantics, so it shouldn't matter one way or the other. `function` is an edge-case since functions themselves don't have reference semantics, but do carry around an environment, which users might expect to be the unique constructor instance evaluation env.

Reply:
I think you don't use  `formals<-`  properly -- you must use  `alist()` : 
```
> a1 <- alist(a=1:10); a2 <- alist(a = quote(1:10)); identical(a1, a2)
[1] FALSE
> f1 <- f2 <- function(){}; formals(f1) <- a1; formals(f2) <- a2; identical(f1,f2)
[1] FALSE
> str(f1)
function (a = 1:10)  
> str(f2)
function (a = quote(1:10))  
> 
```

Reply:
@mmaechler this is an issue with creating functions in non-standard ways, so we definitely want to use `list()` here.

@t-kalinowski while inlining objects into the constructor args directly will generally be safe, I think it's going to potentially cause a lot of confusion.

Reply:
On my first pass at this (unpushed), I did have all branches in `base_default()` wrapped in `quote(...)`. I reverted it once I realized it was unnecessary. 

Also, it's about 5x slower not to inline (I'm not sure how much this matters here; in absolute numbers, these are both very fast).

``` r
f1 <- f2 <- function(x = integer()) x
formals(f2)$x <- integer()

bench::mark(f1(), f2()) |> print() |> plot() |> print()
#> # A tibble: 2 × 13
#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
#> 1 f1()          246ns  328ns  1937884.        0B        0 10000     0     5.16ms
#> 2 f2()           41ns  123ns  5988817.        0B        0 10000     0     1.67ms
#> # ℹ 4 more variables: result <list>, memory <list>, time <list>, gc <list>
#> Loading required namespace: tidyr
```

![](https://i.imgur.com/O1XD8MF.png)<!-- -->

Reply:
I do think that this is more of a user facing documentation issue — i.e. I think we should generally recommend that the user wrap their default values in quote, unless they really know what they're doing.

Reply:
We can use NSE for `new_property(default)` and automatically quote `default`. We could then allow using `I()` as an NSE escape hatch.

That said, as I consider it, I'm coming to the conclusion that using NSE here isn't the best approach. The laziness of a default value should be explicit, clearly communicated to readers, and opt-in for authors.

I agree we'll want to update docs to discuss this and encourage usage of `quote()`.